vcd/resource_vcd_vapp_nat_rules.go

```diff
@@ -0,0 +1,310 @@
+					"level. Useful when connected as sysadmin working across different organizations",
+			},
+			"vdc": {
+				Type:        schema.TypeString,
+				Required:    false,
+				Optional:    true,
+				ForceNew:    true,
+				Description: "The name of VDC to use, optional if defined at provider level",
+			},
+			"vapp_id": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "vApp identifier",
+			},
+			"network_id": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "vApp network identifier",
+			},
+			"nat_type": &schema.Schema{
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validation.StringInSlice([]string{portForwardingNatType, ipTranslationNatType}, false),
+				Description:  "One of: `ipTranslation` (use IP translation), `portForwarding` (use port forwarding).",
+			},
+			"enable_ip_masquerade": &schema.Schema{
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     false,
+				Description: "When enabled translates a virtual machine's private, internal IP address to a public IP address for outbound traffic.",
+			},
+			"rule": &schema.Schema{
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": &schema.Schema{
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: "Id of the rule. Can be used to track syslog messages.",
+						},
+						"mapping_mode": &schema.Schema{
+							Type:         schema.TypeString,
+							Optional:     true,
+							ValidateFunc: validation.StringInSlice([]string{"automatic", "manual"}, false),
+							Description:  "Mapping mode. One of: `automatic`, `manual`",
+						},
+						"vm_id": &schema.Schema{
+							Type:        schema.TypeString,
+							Required:    true,
+							Description: "VM to which this rule applies.",
+						},
+						"vm_nic_id": &schema.Schema{
+							Type:        schema.TypeInt,
+							Required:    true,
+							Description: "VM NIC ID to which this rule applies.",
+						},
+						"external_ip": &schema.Schema{
+							Type:         schema.TypeString,
+							Optional:     true,
+							Computed:     true,
+							ValidateFunc: validation.IsIPAddress,
+							Description:  "External IP address to forward to or External IP address to map to VM",
+						},
+						"external_port": &schema.Schema{
+							Type:        schema.TypeInt,
+							Optional:    true,
+							Description: "External port to forward.",
+						},
+						"forward_to_port": &schema.Schema{
+							Type:        schema.TypeInt,
+							Optional:    true,
+							Description: "Internal port to forward.",
+						},
+						"protocol": &schema.Schema{
+							Type:         schema.TypeString,
+							Optional:     true,
+							ValidateFunc: validation.StringInSlice([]string{"TCP", "UDP", "TCP_UDP"}, false),
```

Comment:
Is "TCP", "UDP", "TCP_UDP" comming from the VCD API?

It's inconsistent with the `vapp_firewall_rules` constants:
> protocol - (Optional) The protocol to match. One of **tcp**, **udp**, icmp, any or **tcp&udp**. Default is any. 

FYI: @Didainius @dataclouder

Reply:
Vaidotas Bauzys wrote:
> Protocol naming from Api

So, we leave as is in the sake of API transparency.